### PR TITLE
MDS-2471 - Tracking progress/stages of a NoW application

### PR DESCRIPTION
**Pull request by mmumby**
opening this PR to get feedback on the db tables/column naming while I continue building the feature

### CODE REVIEW

**Comment by pdittaro** [comment]
I'm conflicted over this where I somewhat feel that the status of the Notice of Work should be a logical reflection of the data rather than a type code. E.g. a NoW application is in Technical Review if it has been imported but there isn't a 'technical review object created'. Or a 'consultation / consultation deferral object'. 

We are going to have to map state to the specific type codes: what happens to existing applications when a type code is set to inactive, or added? 

Can or should this value be realized through a logical column?

**Comment by mmumby** [comment]
The reason I opted for a code table, and an additional `application_progress` table is so we can track: 
- when an application is moved to the next stage - on both sides capturing start_date and end_date
- who moved the application (again, from both sides as)
- the duration the application was in a step (calculated through the UI? or BE and returned to the FE)
- I believe there will be time limits in a future iteration, (ie an application can only be in referral/consultation for 6 weeks)
Other considerations:
- If an application was `rejected` during the technical review, does it skip the `referral/consultation` and go directly to the `decision`
- If an application is moved forward to prematurely, how do we track it moving back a step without overriding the original values


With all that logic, I'm not sure it can be tracked through a single column on the application.
Thoughts? let's chat IRL

File: migrations/sql/V2019.2019.12.29.3.33__add_application_progress.sql [hunk 1]
```diff
@@ -0,0 +1,33 @@
+CREATE TABLE IF NOT EXISTS application_progress_type (
```
Comment (nayr974) [inline migrations/sql/V2019.2019.12.29.3.33__add_application_progress.sql]: I think application_progress_status is more representative, should be a quick change to make. Can you also please add table comments?

Comment (mmumby) [reply migrations/sql/V2019.2019.12.29.3.33__add_application_progress.sql]: Will do!

File: migrations/sql/V2019.2019.12.29.3.33__add_application_progress.sql [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  update_timestamp timestamp with time zone DEFAULT now() NOT NULL
+);
+ALTER TABLE now_application_progress_status OWNER TO mds;
+
+CREATE TABLE IF NOT EXISTS now_application_progress (
+  application_progress_id SERIAL PRIMARY KEY,
+  now_application_id integer,
+  application_progress_status_code character varying(3),
+  active_ind boolean DEFAULT true NOT NULL,
+  end_date character varying(60),
```
[the remark this answers is not in the document]
Comment (mmumby) [reply migrations/sql/V2019.2019.12.29.3.33__add_application_progress.sql]: will remove, this is not needed.

[the remark this answers is not in the document]
Comment (mmumby) [reply migrations/sql/V2019.2019.12.29.3.33__add_application_progress.sql]: not needed. removed

File: services/core-api/app/api/now_applications/models/now_application_progress.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    created_by = db.Column(db.String, nullable=False)
+    active_ind = db.Column(db.Boolean, nullable=False, server_default=FetchedValue())
+
+    def __repr__(self):
+        return '<NOWApplicationProgress %r>' % self.application_progress_id
+
+    @classmethod
+    def create(cls, now_application, application_progress_status_code, add_to_session=True):
+        progress = cls(
+                                                                               # now_application_id=now_application.now_application_id,
```
[the remark this answers is not in the document]
Comment (mmumby) [reply services/core-api/app/api/now_applications/models/now_application_progress.py]: removed

**Pull request closed without merging**